### ISSUE TITLE
Setters for java.util.Optional Properties do not allow null Values

### ISSUE THREAD

**Issue by buett** [immutables/immutables#1166]
Following example:
```
@Value.Immutable
@Value.Style(jdkOnly = true)
public interface TestDTO {
    Optional<String> getName();
}
```
The generated setters in immutable class and builder doesn't allow null values because of using Objects.requireNonNull.
```
public final ImmutableTestDTO withName(String value) {
    @Nullable String newValue = Objects.requireNonNull(value, "name");
    if (Objects.equals(this.name, newValue)) return this;
    return new ImmutableTestDTO(newValue);
  }
```

**Comment by elucash** [comment]
That is correct and is according to design. Use `.withName(Optional.ofNullable(null))` or enable implicit and more prone to errors nulls with `Style.optionalAcceptNullable = true` (then `.withName(null)` would work IIRC. Thanks!

**Comment by buett** [comment]
Thanks, it works!
Is there also a possibility to avoid generating the setters with the optional parameter. Optional as method parameter is a pattern which is often discouraged and IMHO there are good reasons for that. Therefore it would be nice to have a switch to turn it off.

**Comment by elucash** [comment]
> Optional as method parameter is a pattern which is often discouraged and IMHO there are good reasons for that.
pls, don't cite the "propaganda"  B-)   If someone doesn't like optionals, s/he can use `@Nullable` instead, other than that there are no rational reasons why return value position magically different from parameter or field.

**Comment by buett** [comment]
Who decides what propaganda is? But that's not the point. ;-) Optional as result is a good choice for api description and usage.

I only ask if you could provide a way to turn it off for setters, maybe with a Style attribute. 

Automatic property propagation with null values won't work because of method overloading. (Which setter to use?)
And the manually way `.withName(null)` won't work at the moment. It has to be cast explicitly. `.withName((String)null)` isn't a nice way and `.withName(Optional.empty())` doesn't make it better, I think.

**Comment by buett** [comment]
> I only ask if you could provide a way to turn it off for setters, maybe with a Style attribute.

At least for builders. They don't have getters and so the change has no influence to your rule `x.withValue(x.getValue()).equals(x) == true` ;-)

**Comment by Anusien** [comment]
`@Value.Style(optionalAcceptNullable = true)` ?
